In [1]:
import sys
import os

project_root = os.path.abspath('..')
if project_root not in sys.path:
    sys.path.append(project_root)

%load_ext autoreload
%autoreload 2


In [2]:
import pandas as pd
import numpy as np
import polars as pl

import matplotlib.pyplot as plt
import seaborn as sns

from src.data import make_dataset
from src.features import build_features

sns.set_style()
plt.rcParams["figure.figsize"] = (12, 6)

HOSPITAL_A_ANLIZAR = 112103  # Este es el Torax

## Obtencion de Métricas de Egresos Hospitalarios

En este apartado se quieren calculcar diversas métricas relevantes para el análisis de los
egresos hospitalarios en Chile. Las métricas a calcular son las siguientes:

1. Cantidad de Egresos Hospitalarios
2. Dias de Estada Totales
3. Cantidad de Intervenciones Quirúrgicas
4. Cantidad de Muertos

Todas estas métricas serán calculadas para cada Año, Establecimiento de Salud y Diagnóstico
distinto que exista en la base de datos. Por ejemplo:

"En el año 2019, el Hospital San José (código XXXXXX) tuvo XX egresos, con XX dias estada promedio,
con XX intervenciones quirúrgicas y XX muertos para el diagnóstico A24.4"s


In [3]:
df_nacional = pl.scan_csv(
    "../data/processed/egresos_procesados.csv",
    infer_schema_length=10000,
    null_values="Extranjero",
)

In [4]:
AGRUPACION = [
    "ANO_EGRESO",
    "ESTABLECIMIENTO_SALUD",
    "GLOSA_ESTABLECIMIENTO_SALUD",
    "DIAG1",
]

metricas = build_features.obtener_metricas_egresos(df_nacional, AGRUPACION).collect()

In [5]:
display(metricas.sample(10))

shape: (10, 8)
┌────────────┬──────────────┬──────────────┬───────┬───────────┬─────────────┬─────────┬───────────┐
│ ANO_EGRESO ┆ ESTABLECIMIE ┆ GLOSA_ESTABL ┆ DIAG1 ┆ n_egresos ┆ dias_estada ┆ n_int_q ┆ n_muertos │
│ ---        ┆ NTO_SALUD    ┆ ECIMIENTO_SA ┆ ---   ┆ ---       ┆ _totales    ┆ ---     ┆ ---       │
│ i64        ┆ ---          ┆ LUD          ┆ str   ┆ u32       ┆ ---         ┆ i64     ┆ i64       │
│            ┆ i64          ┆ ---          ┆       ┆           ┆ i64         ┆         ┆           │
│            ┆              ┆ str          ┆       ┆           ┆             ┆         ┆           │
╞════════════╪══════════════╪══════════════╪═══════╪═══════════╪═════════════╪═════════╪═══════════╡
│ 2008       ┆ 107107       ┆ Hospital de  ┆ D329  ┆ 1         ┆ 1           ┆ 0       ┆ 0         │
│            ┆              ┆ Petorca      ┆       ┆           ┆             ┆         ┆           │
│ 2007       ┆ 118200       ┆ Hospital     ┆ J304  ┆ 6         ┆ 29          ┆ 0       ┆ 0         │
│            ┆              ┆ Clínico del  ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ Sur S.A.     ┆       ┆           ┆             ┆         ┆           │
│ 2015       ┆ 104103       ┆ Hospital     ┆ L97X  ┆ 1         ┆ 29          ┆ 0       ┆ 0         │
│            ┆              ┆ Provincial   ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ del Huasco   ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ M…           ┆       ┆           ┆             ┆         ┆           │
│ 2011       ┆ 111100       ┆ Hospital     ┆ M410  ┆ 9         ┆ 44          ┆ 2       ┆ 0         │
│            ┆              ┆ Clínico San  ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ Borja-Arria… ┆       ┆           ┆             ┆         ┆           │
│ …          ┆ …            ┆ …            ┆ …     ┆ …         ┆ …           ┆ …       ┆ …         │
│ 2015       ┆ 123100       ┆ Hospital     ┆ K721  ┆ 1         ┆ 1           ┆ 0       ┆ 0         │
│            ┆              ┆ Base San     ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ José de      ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ Osorno       ┆       ┆           ┆             ┆         ┆           │
│ 2017       ┆ 102200       ┆ Clínica      ┆ T819  ┆ 1         ┆ 2           ┆ 0       ┆ 0         │
│            ┆              ┆ Iquique      ┆       ┆           ┆             ┆         ┆           │
│ 2004       ┆ 111221       ┆ Clínica Juan ┆ N63X  ┆ 6         ┆ 7           ┆ 5       ┆ 0         │
│            ┆              ┆ Pablo II     ┆       ┆           ┆             ┆         ┆           │
│ 2013       ┆ 117101       ┆ Hospital     ┆ A153  ┆ 26        ┆ 525         ┆ 1       ┆ 1         │
│            ┆              ┆ Clínico      ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ Herminda     ┆       ┆           ┆             ┆         ┆           │
│            ┆              ┆ Martín…      ┆       ┆           ┆             ┆         ┆           │
└────────────┴──────────────┴──────────────┴───────┴───────────┴─────────────┴─────────┴───────────┘

- Al ver una muestra del calculo realizado, se puede observar la cantidad de egresos, dias de
  estada, numero de intervenciones quirurgicas y cantidad de muertos para 10 duplas de
  hospitales - diagnosticos distintos.


Si bien se tienen las metricas calculadas, ahora se quiere responder a las siguientes preguntas:

- En qué diagnósticos se destaca cada hospital productivamente? Esto significa, en qué diagnósticos
  cada hospital tiene una mayor cantidad de egresos que otros hospitales?

- El hospital tiene un desempeño similar dentro de distintos grupos/estratos de hospitales?
  El hospital se desempeñara de la misma forma nacionalmente, dentro del estrato público, privado,
  etc...?

Para responder a la pregunta se realizará un ranking de todos los hospitales para cada
diagnóstico y cada año observado. Este ranking se realizará para distintos grupos/estratos de
hospitales (Nacional, Hospitales Públicos, Hospitales Privados, Hospitales GRD)


## Análisis de Estratos


In [6]:
dict_estratos = build_features.obtener_diccionario_estratos(df_nacional, HOSPITAL_A_ANLIZAR)

In [7]:
variables_a_rankear = ["n_egresos"]
subgrupo_del_ranking = ["ANO_EGRESO", "DIAG1"]

rankings_nacionales = build_features.agregar_ranking_estratos(
    metricas, dict_estratos, variables_a_rankear, subgrupo_del_ranking
)

rankings_hospital_analizado = rankings_nacionales.filter(
    pl.col("ESTABLECIMIENTO_SALUD") == HOSPITAL_A_ANLIZAR
)

In [8]:
diags_mas_relevantes = rankings_hospital_analizado.filter(
    (pl.col("ranking_nacionales_n_egresos") == 1)
    & (pl.col("n_int_q") > 0)
    & (pl.col("ANO_EGRESO") > 2011)
).sort(by=["ANO_EGRESO", "n_egresos", "n_int_q"], descending=True)

veinte_mas_relevantes_por_anio = diags_mas_relevantes.groupby(["ANO_EGRESO"]).head(20)

C:\Users\ppizarro\AppData\Local\Temp\ipykernel_2160\516069765.py:7: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  veinte_mas_relevantes_por_anio = diags_mas_relevantes.groupby(["ANO_EGRESO"]).head(20)


In [9]:
DIAGS_MAS_RELEVANTES = veinte_mas_relevantes_por_anio.select(pl.col("DIAG1").unique()).sort(
    by="DIAG1"
)

In [10]:
print(DIAGS_MAS_RELEVANTES)

shape: (46, 1)
┌───────┐
│ DIAG1 │
│ ---   │
│ str   │
╞═══════╡
│ C33X  │
│ C340  │
│ C341  │
│ C342  │
│ …     │
│ Q231  │
│ Q676  │
│ T820  │
│ Z450  │
└───────┘


Para hacer un contorl de lo que se calculo, se observará una muestra de 10 hospitales - diagnósticos
distintos


In [11]:
display(rankings_nacionales.sample(10))

shape: (10, 27)
┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐
│ ANO_EGRES ┆ ESTABLECI ┆ GLOSA_EST ┆ Capítulo  ┆ … ┆ %_grd_n_e ┆ ranking_i ┆ total_int ┆ %_intern │
│ O         ┆ MIENTO_SA ┆ ABLECIMIE ┆ ---       ┆   ┆ gresos    ┆ nterno_n_ ┆ erno_n_eg ┆ o_n_egre │
│ ---       ┆ LUD       ┆ NTO_SALUD ┆ str       ┆   ┆ ---       ┆ egresos   ┆ resos     ┆ sos      │
│ i64       ┆ ---       ┆ ---       ┆           ┆   ┆ f64       ┆ ---       ┆ ---       ┆ ---      │
│           ┆ i64       ┆ str       ┆           ┆   ┆           ┆ u32       ┆ u32       ┆ f64      │
╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡
│ 2001      ┆ 110100    ┆ Hospital  ┆ Cap.19    ┆ … ┆ 0.011905  ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ San Juan  ┆ LESIONES  ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ de Dios   ┆ TRAUMÁTIC ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ (Santi…   ┆ AS, EN…   ┆   ┆           ┆           ┆           ┆          │
│ 2015      ┆ 114223    ┆ Clínica   ┆ Cap.13    ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Vespucio  ┆ ENFERMEDA ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆           ┆ DES DEL   ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆           ┆ APARATO…  ┆   ┆           ┆           ┆           ┆          │
│ 2014      ┆ 113130    ┆ Hospital  ┆ Cap.19    ┆ … ┆ 0.012048  ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Dr.       ┆ LESIONES  ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Exequiel  ┆ TRAUMÁTIC ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ González  ┆ AS, EN…   ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ C…        ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2006      ┆ 105208    ┆ Clínica   ┆ Cap.19    ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Regional  ┆ LESIONES  ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Elqui     ┆ TRAUMÁTIC ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆           ┆ AS, EN…   ┆   ┆           ┆           ┆           ┆          │
│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │
│ 2001      ┆ 107100    ┆ Hospital  ┆ Cap.09    ┆ … ┆ 0.020833  ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Dr.       ┆ ENFERMEDA ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Gustavo   ┆ DES DEL   ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Fricke    ┆ APARATO…  ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ (Viñ…     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2008      ┆ 104203    ┆ Centro de ┆ Cap.19    ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Salud     ┆ LESIONES  ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Mutual    ┆ TRAUMÁTIC ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ CChC      ┆ AS, EN…   ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆ Copi…     ┆           ┆   ┆           ┆           ┆           ┆          │
│ 2002      ┆ 109201    ┆ Clínica   ┆ Cap.13    ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │
│           ┆           ┆ Dávila    ┆ ENFERMEDA ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆           ┆ DES DEL   ┆   ┆           ┆           ┆           ┆          │
│           ┆           ┆           ┆ APARATO…  ┆   ┆           ┆           ┆           ┆          │
│ 2019      ┆ 112101    ┆ Hospital  ┆ Cap.02 

En la muestra anterior se puede observar las métricas previamente calculada, y el ranking
respectivo que tiene cada hospital para diagnósticos en cada año específico.


In [12]:
rankings_nacionales.to_pandas().to_csv(
    "../data/interim/ranking_nacional_egresos.csv",
    sep=";",
    decimal=".",
    encoding="latin-1",
    index=False,
)

rankings_hospital_analizado.to_pandas().to_csv(
    "../data/interim/ranking_torax_egresos.csv",
    sep=";",
    decimal=".",
    encoding="latin-1",
    index=False,
)

## Metricas para personas sobre 14 años

En este análisis se quiere ver el ranking de egresos pero solamente tomando en cuenta a personas
mayores a 14 años. Esto, para dilucidar si el hospital del tórax es altamente relevante para tratar
patologías congénitas de adultos.


In [14]:
df_nacional_mayores_a_14 = df_nacional.filter(pl.col("EDAD_A_OS") > 14)

In [15]:
metricas_sobre_14 = build_features.obtener_metricas_egresos(
    df_nacional_mayores_a_14, AGRUPACION
).collect()

In [16]:
variables_a_rankear = ["n_egresos"]
subgrupo_del_ranking = ["ANO_EGRESO", "DIAG1"]

rankings_nacionales_sobre_14 = build_features.agregar_ranking_estratos(
    metricas_sobre_14, dict_estratos, variables_a_rankear, subgrupo_del_ranking
)

diags_congenitos_relevantes = rankings_nacionales_sobre_14.filter(
    (pl.col("DIAG1").is_in(["Q230", "Q211", "Q231"])) & (pl.col("ANO_EGRESO") == 2019)
)

In [17]:
diags_congenitos_relevantes.to_pandas().to_csv(
    "../data/interim/diags_congenitas_sobre_14_egresos.csv",
    sep=";",
    decimal=".",
    encoding="latin-1",
    index=False,
)

## 3. Análisis de procedimientos DEIS

En este análisis se quieren responder las siguientes preguntas:

1. ¿Cuál es el procedimiento más frecuente dentro de los pacientes que se les realizó un procedimiento
   quirúrgico?


In [18]:
proced = (
    df_nacional.filter(pl.col("INTERV_Q") == 1)
    .groupby(["ANO_EGRESO", "ESTABLECIMIENTO_SALUD", "GLOSA_PROCED_PPAL"])
    .agg(pl.count())
).collect()

C:\Users\ppizarro\AppData\Local\Temp\ipykernel_2160\3397463805.py:3: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  .groupby(["ANO_EGRESO", "ESTABLECIMIENTO_SALUD", "GLOSA_PROCED_PPAL"])


In [19]:
display(proced)

shape: (10_757, 4)
┌────────────┬───────────────────────┬───────────────────────────────────┬───────┐
│ ANO_EGRESO ┆ ESTABLECIMIENTO_SALUD ┆ GLOSA_PROCED_PPAL                 ┆ count │
│ ---        ┆ ---                   ┆ ---                               ┆ ---   │
│ i64        ┆ i64                   ┆ str                               ┆ u32   │
╞════════════╪═══════════════════════╪═══════════════════════════════════╪═══════╡
│ 2001       ┆ 103203                ┆ null                              ┆ 5921  │
│ 2001       ┆ 107212                ┆ null                              ┆ 1504  │
│ 2001       ┆ 106100                ┆ null                              ┆ 9333  │
│ 2001       ┆ 107100                ┆ null                              ┆ 9959  │
│ …          ┆ …                     ┆ …                                 ┆ …     │
│ 2020       ┆ 122202                ┆ Cistoscopia y/o uretrocistoscopi… ┆ 1     │
│ 2020       ┆ 112530                ┆ Radiografía de esófago, estómago… ┆ 1     │
│ 2020       ┆ 111100                ┆ Ano-recto-sigmoidoscopia en adul… ┆ 1     │
│ 2020       ┆ 133150                ┆ Trombolisis arterial periferica   ┆ 1     │
└────────────┴───────────────────────┴───────────────────────────────────┴───────┘

Los resultados muestran el desglose por año y recinto hospitalario junto a la cantidad de
procedimientos realizados en cada intervención quirúrgica. Sin embargo, se puede observar que
existe una cantidad importante de casos donde la glosa del procedimiento fue nula.

Ahora, se quiere calcular cuántas intervenciones quirúrgicas carecen de una glosa de procedimiento.


In [20]:
display(proced.groupby("GLOSA_PROCED_PPAL").agg(pl.col("count").sum()).sort("count"))

C:\Users\ppizarro\AppData\Local\Temp\ipykernel_2160\989951252.py:1: DeprecationWarning: `groupby` is deprecated. It has been renamed to `group_by`.
  display(proced.groupby("GLOSA_PROCED_PPAL").agg(pl.col("count").sum()).sort("count"))


shape: (422, 2)
┌───────────────────────────────────┬──────────┐
│ GLOSA_PROCED_PPAL                 ┆ count    │
│ ---                               ┆ ---      │
│ str                               ┆ u32      │
╞═══════════════════════════════════╪══════════╡
│ Criocoagulacion conjuntival, cor… ┆ 1        │
│ Test de relaciones objetales      ┆ 1        │
│ Rinomanometria c/s vasocontricto… ┆ 1        │
│ E.E.G. Post-privacion de sueno (… ┆ 1        │
│ …                                 ┆ …        │
│ Instilacion vesical (incluye col… ┆ 7685     │
│ Código no existe actualmente      ┆ 11155    │
│ Apoyo fluoroscópico a procedimie… ┆ 15775    │
│ null                              ┆ 12382642 │
└───────────────────────────────────┴──────────┘

Por lo tanto, 12382642 de todas las intervenciones quirúrgicas carecen de una glosa de procedimiento.


In [21]:
proced.write_excel("../data/interim/conteo_procedimientos.xlsx")